plugin/src/main/java/org/opennms/timeseries/impl/influxdb/InfluxdbWriter.java

```diff
@@ -0,0 +1,166 @@
+        Dispatcher dispatcher = new Dispatcher();
+        dispatcher.setMaxRequests(config.getMaxConcurrentHttpConnections());
+        dispatcher.setMaxRequestsPerHost(config.getMaxConcurrentHttpConnections());
+
+        this.client = new OkHttpClient.Builder()
+                .readTimeout(config.getReadTimeoutInMs(), TimeUnit.MILLISECONDS)
+                .writeTimeout(config.getWriteTimeoutInMs(), TimeUnit.MILLISECONDS)
+                .dispatcher(dispatcher)
+                .connectionPool(connectionPool)
+                .build();
+
+
+        BulkheadConfig bulkheadConfig = BulkheadConfig.custom()
+                .maxConcurrentCalls(config.getMaxConcurrentHttpConnections() * 6)
+                .maxWaitDuration(Duration.ofMillis(config.getBulkheadMaxWaitDurationInMs()))
+                .fairCallHandlingStrategyEnabled(true)
+                .build();
+        asyncHttpCallsBulkhead = Bulkhead.of("asyncHttpCalls", bulkheadConfig);
+
+        // Expose HTTP client statistics
+        metrics.register("connectionCount", (Gauge<Integer>) () -> client.connectionPool().connectionCount());
+        metrics.register("idleConnectionCount", (Gauge<Integer>) () -> client.connectionPool().idleConnectionCount());
+        metrics.register("queuedCallsCount", (Gauge<Integer>) () -> client.dispatcher().queuedCallsCount());
+        metrics.register("runningCallsCount", (Gauge<Integer>) () -> client.dispatcher().runningCallsCount());
+        metrics.register("availableConcurrentCalls", (Gauge<Integer>) () -> asyncHttpCallsBulkhead.getMetrics().getAvailableConcurrentCalls());
+        metrics.register("maxAllowedConcurrentCalls", (Gauge<Integer>) () -> asyncHttpCallsBulkhead.getMetrics().getMaxAllowedConcurrentCalls());
+    }
+
+    public void writePoints(final List<Point> points) {
+        asyncHttpCallsBulkhead.executeCompletionStage(() -> writeAsync(points)).whenComplete((r, ex) -> {
+            if (ex == null) {
+                samplesWritten.mark(points.size());
+            } else {
+                // FIXME: Data loss
+                samplesLost.mark(points.size());
+                LOG.error("Error occurred while storing samples, sample will be lost.", ex);
+            }
+        });
+    }
+
+    public CompletableFuture<Void> writeAsync(List<Point> points) {
+        String bodyContent = points
+                .stream()
+                .map(Point::toLineProtocol)
+                .collect(Collectors.joining("\n"));
+
+        final CompletableFuture<Void> future = new CompletableFuture<>();
+
+        final byte[] compressed;
+        try {
+            compressed = compress(bodyContent);
+        } catch (IOException e) {
+            future.completeExceptionally(e);
+            return future;
+        }
+        final RequestBody body = RequestBody.create(compressed);
+
+
+        // TODO: Patrick make URL composition nicer
+        String url = config.getUrl()
```

Comment:
We should built this once instead of computing it for every call. `HttpUrl.Builder` can help: https://square.github.io/okhttp/4.x/okhttp/okhttp3/-http-url/

Reply:
> We should built this once instead of computing it for every call.
fixed

> HttpUrl.Builder can help: https://square.github.io/okhttp/4.x/okhttp/okhttp3/-http-url/
done (although it doesn't look too elegant either)